### PR TITLE
Added multibanco payment method

### PR DESCRIPTION
**Pull request by lfluvisotto**
**Description**
Added multibanco payment method

**Tested scenarios**
Make a purchase using the multibanco payment method

**Fixed issue**: 
N/A

### CODE REVIEW

**Comment by lfluvisotto** [comment]
@rikterbeek

**Comment by rikterbeek** [comment]
Hi @lfluvisotto,

Many thanks for the PR! We will investigate this and test it and get back to you, but it looks great.

Regards,
Rik
Adyen

File: app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            $isAvailable = false;
+        }
+
+        return $isAvailable;
+    }
+
+    /**
+     * @return bool
+     */
+    public function canCreateAdyenSubscription()
```
Comment (rikterbeek) [inline app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php]: This is not allowed for subscription could you remove this methods

File: app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isBillingAgreement()
```
Comment (rikterbeek) [inline app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php]: This is not allowed for subscription could you remove this methods

File: app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        return true;
+    }
+
+    /**
+     * @param Adyen_Payment_Model_Billing_Agreement $billingAgreement
+     * @param array $data
+     *
+     * @return $this
+     */
+    public function parseRecurringContractData(Adyen_Payment_Model_Billing_Agreement $billingAgreement, array $data)
```
Comment (rikterbeek) [inline app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php]: This is not allowed for subscription could you remove this methods

File: app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        return $this;
+    }
+
+    /**
+     * @param Adyen_Payment_Model_Billing_Agreement $billingAgreement
+     * @param Mage_Sales_Model_Quote_Payment $paymentInfo
+     *
+     * @return $this
+     */
+    public function initBillingAgreementPaymentInfo(Adyen_Payment_Model_Billing_Agreement $billingAgreement, Mage_Sales_Model_Quote_Payment $paymentInfo)
```
Comment (rikterbeek) [inline app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php]: This is not allowed for subscription could you remove this methods

File: app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        $this->_code = $code;
+        return $this;
+    }
+
+    /**
+     * @desc CustomerInteraction is set by the recurring_payment_type or controlled by Adyen_Subscription module
+     * @param $customerInteraction
+     */
+    public function setCustomerInteraction($customerInteraction)
```
Comment (rikterbeek) [inline app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php]: This method is only for the oneclick method and can be removed

File: app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        $this->_customerInteraction = (bool)$customerInteraction;
+    }
+
+    /**
+     * @return Adyen_Payment_Model_Billing_Agreement
+     */
+    public function getBillingAgreement()
```
Comment (rikterbeek) [inline app/code/community/Adyen/Payment/Model/Adyen/Multibanco.php]: This is not allowed for subscription could you remove this methods

**Comment by rikterbeek** [comment]
Hi @lfluvisotto,

The code looks really good thank you. A couple of methods that are not needed because it does not support recurring/subscription. Can you remove these methods see above. As well Multibance comes back as well as HPP method. Could you add the check here (example SEPA direct debet see: https://github.com/Adyen/adyen-magento/blob/develop/app/code/community/Adyen/Payment/Model/Observer.php#L149-L152) to disable it when the API method is enabled?

Regards,
Rik
Adyen

**Comment by rikterbeek** [comment]
Hi @lfluvisotto  

Yes indeed perfect!

Regards,
Rik
Adyen

**Comment by lancergr** [comment]
Hey @lfluvisotto 

Thanks for adding multibanco - the PR looks fine!

One minor improvement can be that instead of calculating the comprafacil.deadline_date two times (https://github.com/Adyen/adyen-magento/pull/856/files#diff-857809b571a4129f6be9571f298a001cR89 and https://github.com/Adyen/adyen-magento/pull/856/files#diff-12fb05611e5c9a93cea6b697bc051775R23)
you can calculate it once and store it on additional data, here:
https://github.com/Adyen/adyen-magento/pull/856/files#diff-9f51100419c8e5c7060aa2eb0295c381R569
or expose the calculation on a helper

Please let me know if you can include it to the PR.


Best Regards,
Giorgos
Adyen

**Comment by lfluvisotto** [comment]
@lancergr yes, I will verify.

**Comment by lfluvisotto** [comment]
@lancergr can it be this way ?
https://github.com/Adyen/adyen-magento/pull/856/commits/fa832e3bf48eb6b149813bef844f38470bc214dc
https://github.com/Adyen/adyen-magento/pull/856/commits/cc2081cf32cca1dcaddc2fdb9b5af4f1640b39b0
https://github.com/Adyen/adyen-magento/pull/856/commits/207eb60aa2b280e0db2b0f8bb89c41630daf26a6

**Comment by lancergr** [comment]
Hi @lfluvisotto 

Exactly!
It's also good practice to avoid nested ifs when possible. Could you please also move the if statement of line 572 https://github.com/Adyen/adyen-magento/commit/fa832e3bf48eb6b149813bef844f38470bc214dc#diff-9f51100419c8e5c7060aa2eb0295c381R572
outside the if of line 569?
https://github.com/Adyen/adyen-magento/commit/fa832e3bf48eb6b149813bef844f38470bc214dc#diff-9f51100419c8e5c7060aa2eb0295c381R569

Best Regards,
Giorgos Adyen

**Comment by lfluvisotto** [comment]
@lancergr https://github.com/Adyen/adyen-magento/pull/856/commits/3beb2abe56bfc68c61957b90038a3faddac02649

**Comment by rikterbeek** [comment]
Hi @lfluvisotto,

Many thanks for the PR and updates. I will merge it now.

Regards,
Rik
Adyen

**Pull request merged**